lib/media/media_source_engine.js

```diff
@@ -2451,91 +2464,92 @@ shaka.media.MediaSourceEngine = class {
-      await this.changeType(contentType, newMimeType, transmuxer);
+      return {
+        type: 'changeType',
+        newMimeType,
+        transmuxer,
+      };
     } else {
       if (transmuxer) {
         transmuxer.destroy();
       }
-      await this.reset(streamsByType);
+      return {
+        type: 'reset',
+        newMimeType,
+        transmuxer: null,
+      };
     }
-    return true;
   }
 
   /**
-   * Returns true if it's necessary codec switch to load the new stream.
-   *
+   * Codec switch if necessary, this will not resolve until the codec
+   * switch is over.
    * @param {shaka.util.ManifestParserUtils.ContentType} contentType
-   * @param {string} refMimeType
-   * @param {string} refCodecs
-   * @return {boolean}
+   * @param {string} mimeType
+   * @param {string} codecs
+   * @param {!Map<shaka.util.ManifestParserUtils.ContentType,
+   *               shaka.extern.Stream>} streamsByType
+   * @return {!Promise<boolean>} true if there was a codec switch,
+   *                              false otherwise.
    * @private
    */
-  isCodecSwitchNecessary_(contentType, refMimeType, refCodecs) {
-    if (contentType == shaka.util.ManifestParserUtils.ContentType.TEXT) {
-      return false;
+  async codecSwitchIfNecessary_(contentType, mimeType, codecs, streamsByType) {
+    const ContentType = shaka.util.ManifestParserUtils.ContentType;
+    const {type, transmuxer, newMimeType} = this.getInfoAboutResetOrChangeType_(
+        contentType, mimeType, codecs, streamsByType);
+
+    const newAudioStream = streamsByType.get(ContentType.AUDIO);
+    if (newAudioStream) {
+      this.needSplitMuxedContent_ = newAudioStream.isAudioMuxedInVideo;
     }
-    const MimeUtils = shaka.util.MimeUtils;
-    const currentCodec = MimeUtils.getNormalizedCodec(
-        MimeUtils.getCodecs(this.sourceBufferTypes_.get(contentType)));
-    const currentBasicType = MimeUtils.getBasicType(
-        this.sourceBufferTypes_.get(contentType));
 
-    let newMimeType = shaka.util.MimeUtils.getFullType(refMimeType, refCodecs);
-    let needTransmux = this.config_.forceTransmux;
-    if (!shaka.media.Capabilities.isTypeSupported(newMimeType) ||
-        (!this.sequenceMode_ &&
-        shaka.util.MimeUtils.RAW_FORMATS.includes(newMimeType))) {
-      needTransmux = true;
+    if (type == 'none') {
+      if (this.transmuxers_.has(contentType) && !transmuxer) {
+        this.transmuxers_.get(contentType).destroy();
+        this.transmuxers_.delete(contentType);
+      } else if (!this.transmuxers_.has(contentType) && transmuxer) {
+        this.transmuxers_.set(contentType, transmuxer);
+      } else if (transmuxer) {
+        // Compare if the transmuxer is different
+        if (this.transmuxers_.has(contentType) &&
+            this.transmuxers_.get(contentType).transmux !=
+              transmuxer.transmux) {
+          this.transmuxers_.get(contentType).destroy();
+          this.transmuxers_.delete(contentType);
```

Comment:
No need to delete, as set will overwrite it anyway.

Reply:
Done!